tests/std/tests/P0019R8_atomic_ref/test.cpp

```diff
@@ -34,50 +34,191 @@ struct int128 {
+    }
+
+    if constexpr (std::is_integral_v<TD> && !std::is_same_v<bool, TD> && !std::is_const_v<T>) {
+        static_assert(requires(const AR& r, TD v, std::memory_order ord) {
+            { r.fetch_and(v) } -> std::same_as<TD>;
+            { r.fetch_and(v, ord) } -> std::same_as<TD>;
+            { r.fetch_or(v) } -> std::same_as<TD>;
+            { r.fetch_or(v, ord) } -> std::same_as<TD>;
+            { r.fetch_xor(v) } -> std::same_as<TD>;
+            { r.fetch_xor(v, ord) } -> std::same_as<TD>;
+            { r &= v } -> std::same_as<TD>;
+            { r |= v } -> std::same_as<TD>;
+            { r ^= v } -> std::same_as<TD>;
+        });
+    } else {
+        static_assert(!requires(const AR& r, TD v) { r.fetch_and(v); });
+        static_assert(!requires(const AR& r, TD v, std::memory_order ord) { r.fetch_and(v, ord); });
+        static_assert(!requires(const AR& r, TD v) { r.fetch_or(v); });
+        static_assert(!requires(const AR& r, TD v, std::memory_order ord) { r.fetch_or(v, ord); });
+        static_assert(!requires(const AR& r, TD v) { r.fetch_xor(v); });
+        static_assert(!requires(const AR& r, TD v, std::memory_order ord) { r.fetch_xor(v, ord); });
+        static_assert(!requires(const AR& r, TD v) { r &= v; });
+        static_assert(!requires(const AR& r, TD v) { r |= v; });
+        static_assert(!requires(const AR& r, TD v) { r ^= v; });
+    }
+}
+
 template <class T>
-constexpr bool atomic_ref_has_member_difference_type = requires { typename std::atomic_ref<T>::difference_type; };
-
-static_assert(std::is_same_v<std::atomic_ref<signed char>::difference_type, signed char>);
-static_assert(std::is_same_v<std::atomic_ref<short>::difference_type, short>);
-static_assert(std::is_same_v<std::atomic_ref<int>::difference_type, int>);
-static_assert(std::is_same_v<std::atomic_ref<long>::difference_type, long>);
-static_assert(std::is_same_v<std::atomic_ref<long long>::difference_type, long long>);
-static_assert(std::is_same_v<std::atomic_ref<unsigned char>::difference_type, unsigned char>);
-static_assert(std::is_same_v<std::atomic_ref<unsigned short>::difference_type, unsigned short>);
-static_assert(std::is_same_v<std::atomic_ref<unsigned int>::difference_type, unsigned int>);
-static_assert(std::is_same_v<std::atomic_ref<unsigned long>::difference_type, unsigned long>);
-static_assert(std::is_same_v<std::atomic_ref<unsigned long long>::difference_type, unsigned long long>);
-static_assert(std::is_same_v<std::atomic_ref<char>::difference_type, char>);
+void test_atomic_ref_constraints_cv() { // COMPILE-ONLY
+    static_assert(!std::is_const_v<T> && !std::is_volatile_v<T>);
+    test_atomic_ref_constraints_single<T>();
+    test_atomic_ref_constraints_single<const T>();
+    if constexpr (std::atomic_ref<T>::is_always_lock_free) {
+        test_atomic_ref_constraints_single<volatile T>();
+        test_atomic_ref_constraints_single<const volatile T>();
+    }
+}

```

Comment:
The currently test coverage unfortunately skips instantiation, and the instantiation of `std::atomic_ref<volatile int>::wait` still fails. More changes are needed.